src/librustc_middle/mir/cache.rs

```diff
@@ -197,14 +197,18 @@ pub struct ReadOnlyBodyAndCache<'a, 'tcx> {
 }
 
 impl ReadOnlyBodyAndCache<'a, 'tcx> {
-    fn new(body: &'a Body<'tcx>, cache: &'a Cache) -> Self {
+    pub fn new(body: &'a Body<'tcx>, cache: &'a Cache) -> Self {
         assert!(
             cache.predecessors.is_some(),
             "Cannot construct ReadOnlyBodyAndCache without computed predecessors"
         );
         Self { body, cache }
     }
 
+    pub fn new_unchecked(body: &'a Body<'tcx>, cache: &'a Cache) -> Self {
```

Comment:
This method doesn't have any documentation, so I cannot tell if it is meant to be an efficiency hack (which seems very unlikely, since an assert of `is_some()` should be incredibly cheap to check, no), or if it is something where at the call site (or at future expected call sites) you cannot guarantee that `cache.predecessors.is_some()`.

(I briefly tried to infer the intent by looking at the diff, so that I might suggest what comment to write here, or whether to advise removing the method entirely... From what I can tell, at the single call site where this is called, you are invoking `cache.ensure_predecessors` before this is invoked for the backwards flow case, so that will ensure that `cache.predecessors.is_some()`. But I'm guessing that predecessors is none for the forwards flow case...)

This is all a long winded way of saying that you should spell out in a doc-comment under what circumstances you expect `ReadOnlyBodyAndCache::new_unchecked` to be invoked rather than `ReadOnlyBodyAndCache::new`

Reply:
This is the part that will be removed after #71044 lands. Sorry. That should have been in the commit message for the `HACK` commit.

This is here because we want `join_state_into_successors` to have the same signature for both forward and backward analyses. Only backward analyses need to access the predecessor graph, but the current API demands that you must compute and cache the predecessor graph before creating a `ReadOnlyBodyAndCache` even if you don't actually use it.

FWIW, figuring this out how this was supposed to work caused me a great deal of consternation and was the motivation for #71044, which goes back to using `Body` everywhere.

Reply:
Maybe we should fold #71044 into this PR? I'll finish looking at this PR first before I look at that one.

Reply:
Oh but perhaps you are concerned about performance issues with PR #71044, while you are confident that this PR should not inject performance problems, and thus it would be better to land them separately; is that a correct inference?

Reply:
To be clear, I was also concerned about regressions from this PR. I wanted to benchmark #71044 separately from this one, since the concerns/potential performance issues are very different in that PR (additional locking and slow serialization) than this one (renaming APIs and additional layers of abstraction).

Reply:
I've rebased on the latest master, which includes #71044. The HACK commit has been removed. Now both forward and backward dataflow analyses can use `Body` everywhere, and the reverse CFG will only be computed when necessary (see the most recent commit).